### PR TITLE
VAULT-8631 Make upgrade synchronous when no keys to upgrade

### PR DESCRIPTION
**Pull request by VioletHynes**
This makes the upgrade synchronous when there are no keys in the KV mount (such as when we are creating a new KVV2 mount).

This should make things easier as the mount will be available to use as soon as the mount creation request succeeds, making things easier for users.

All tests pass (and are much faster!) and all tests should implicitly test this feature. Introducing some latency into the upgrade process make the tests fail consistently without my change and pass consistently with it.

### CODE REVIEW

File: upgrade.go [hunk 1]
```diff
@@ -118,6 +118,18 @@ func (b *versionedKVBackend) Upgrade(ctx context.Context, s logical.Storage) err
 		return nil
 	}
 
+	// If we have 0 keys, it's either a new mount or one that's trivial to upgrade,
+	// so we should do the upgrade synchronously
+	upgradeSynchronously := false
+	keys, err := logical.CollectKeys(ctx, s)
```
Comment (ncabatoff) [inline upgrade.go:124]: This is probably OK, but it'd be nice if we could, instead of reading everything for this mount, stop as soon as we found at least one key.  I suppose you could do that using ScanView with a callback that cancels the context passed into ScanView, after flagging the presence of keys as a side-effect.

Comment (ncabatoff) [reply upgrade.go:124]: I don't think Done() ends a context, it's to check if a context is done.  You'd need to use context.WithCancel and then call the cancel func.

Comment (VioletHynes) [reply upgrade.go:124]: Right, of course. 

This approach seems to result in errors. As a result, I'm not sure there's a way to gracefully exit a `ScanView` by cancelling the context. In particular, I get:
```
2022-09-29T11:37:57.788-0400 [ERROR] upgrading resulted in error: error="list failed at path \"999998/\": context canceled"
    upgrade_test.go:51: list failed at path "999998/": context canceled
--- FAIL: TestVersionedKV_Upgrade (2.38s)
```

Comment (ncabatoff) [reply upgrade.go:124]: Yes, it will appear as an error.  But you can first just look at the side-effect of the callback.  If it reported keys found, ignore the context cancelled error.

Comment (VioletHynes) [reply upgrade.go:124]: Yeah, I mean, I could change the check to the following:
```
	if err != nil && (!strings.Contains(err.Error(), "context canceled"))
```

I guess this doesn't feel particularly graceful to me either. I'm willing to believe either approach is the lesser of two evils, but I guess it's the classic trade-off between code cleanliness and speed.

What do you think?

Comment (VioletHynes) [reply upgrade.go:124]: Right - this was your initial suggestion, after all. I feel uneasy about not checking an error in a certain case, even if we think it "shouldn't" have any other errors than the cancel error.

Comment (ncabatoff) [reply upgrade.go:124]: Well, it's ok to check it as you were proposing.  I prefer speed over code simplicity in this case.

Comment (VioletHynes) [reply upgrade.go:124]: Yeah, I think I do favour the original implementation in this case. The speed difference should be the same for the vast majority of cases (new KVV2 mount), and upgrades are affected only a tiny amount (and in a way that should be completely negligible compared to the cost of upgrading millions of keys), so this would be a tiny operation in the cases where it's actually happening.

If this increased speed for new KVV2 case, I think I'd go with the speed approach no question.

All this being said, it's not a hill I'm willing to die on and I do like both approaches here.

Comment (VioletHynes) [reply upgrade.go:124]: The diff of changes would probably be this:
```
	upgradeSynchronously := true
	scanContext, cancelFunc := context.WithCancel(context.Background())
	callback := func(path string) {
		// We've found a key, so break.
		upgradeSynchronously = false
		cancelFunc()
	}
	err := logical.ScanView(scanContext, logical.NewStorageView(s, ""), callback)
	if err != nil && (!strings.Contains(err.Error(), "context canceled")) {
		b.Logger().Error("upgrading resulted in error", "error", err)
		return err
	}
```

Thanks for raising this, regardless. I hadn't yet given enough thought to this alternate approach.

Comment (VioletHynes) [reply upgrade.go:124]: I considered that approach first -- I moved away from it as there'd be nothing stopping the same arg to a tune and doing a synchronous upgrade.

Today, an upgrade is performed by the following body sent to the tune endpoint (the same body is sent when asking for a new KVV2, also):
```
{
  "options": {
      "version": "2"
  }
}
```

We could add another option, like "synchronous" or "upgrade" (where if "upgrade" is set to false, we go synchronous).

The reason I don't like this is that we'd have (discoverable via OSS repos but) undocumented options _or_ we'd need to document when to use this option. I think the chance for (minor) foot-guns is high enough here, because we can't really give an estimate for when it'll be done and people may underestimate the time required.

To put it another way - there's no way that I could see to enable this kind of communication when creating a mount without adding the option when tuning, and I think that, in general, it's probably rarely a good idea for an existing KV mount in reality. 

I don't know that the added time is particularly significant here, even for many keys. The bulk of the time spent is writing new versions of, then deleting the old keys.

Comment (VioletHynes) [reply upgrade.go:124]: For the record, I don't know if we've precedent for undocumented options like this upon mount creation, and if so, I might be making a mountain out of a molehill, but at the very least, it's probably useful to describe the reasons I took this path.

Comment (ccapurso) [reply upgrade.go:124]: That's fair. I am not sure how long it my take to perform the `CollectKeys` so I was mainly concerned that it might cause upgrades for KVv1 to be much slower.

**Review by ncabatoff — APPROVED** [review]
Excited to finally have this!

File: upgrade.go (hunk 1 above)
Comment (ccapurso) [reply upgrade.go:124]: Great, sounds good to me!

**Review by ccapurso — APPROVED** [review]
Looks good, will be great to have this change!
